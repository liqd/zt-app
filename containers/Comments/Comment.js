import React, { useState, useCallback, useEffect } from 'react';
import { Alert, View, Image } from 'react-native';
import { Button } from '@rneui/base';
import { styles } from './Comment.styles';
import { TextSourceSans } from '../../components/TextSourceSans';
import IconSLI from 'react-native-vector-icons/SimpleLineIcons';
import { DateService } from '../../services/DateService';
import { TouchableWithoutFeedback } from 'react-native';
import { ButtonCounter } from '../../components/ButtonCounter';
import { SubComments } from './SubComments';
import API from '../../BaseApi';
import AsyncStorage from '@react-native-async-storage/async-storage';

const NUM_OF_LINES = 2;

export const Comment = (props) => {
  const [showSubComments, setShowSubComments] = useState(false);
  const [showWholeComment, setShowWholeComment] = useState(false);
  const [hasExcerpt, setHasExcerpt] = useState(false);
  const [processing, setProcessing] = useState(false);
  const [comment, setComment] = useState(props.comment);
  const [commentBeingProcessed, setCommentBeingProcessed] = useState(props.comment);
  const commentMenuItems = [
    {
      title: 'Edit',
      icon: 'pencil',
      action: () => console.log('Edit comment'),
      isFirst: true,
      isAllowed: comment.has_changing_permission
    },
    {
      title: 'Delete',
      icon: 'trash',
      action: () => props.toggleDeleteModal(),
      isAllowed: comment.has_deleting_permission
    },
    {
      title: 'Report',
      icon: 'flag',
      action: () => console.log('Report comment'),
      isFirst: !comment.has_changing_permission && !comment.has_deleting_permission,
      isLast: true,
      isAllowed: true
    },
    {
      title: 'Cancel',
      action: () => props.toggleMenu(),
      isCancel: true,
      isAllowed: true
    },
  ];

  const commentDeleteModalItems = [
    {
      // space is to center the text
      title: '   This comment will be deleted.\nThis action cannot be undone.',
      isText: true
    },
    {
      title: 'Delete',
      action: () => deleteComment()
    },
    {
      title: 'Cancel',
      action: () => props.toggleDeleteModal(),
      isCancel: true
    },
  ];

  useEffect(() => {
    setComment(props.comment);
    props.openSubComments && setShowSubComments(true);
  }, [props.openSubComments, props.comment]);

  useEffect(() => {
    props.setDeleteModalItems(commentDeleteModalItems);
    props.setMenuItems(commentMenuItems);
  }, [commentBeingProcessed]);

  const toggleSubComments = () => {
    setShowSubComments(!showSubComments);
  };

  const toggleWholeComment = () => {
    setShowWholeComment(!showWholeComment);
  };

  const onTextLayout = useCallback(e => {
    setHasExcerpt(e.nativeEvent.lines.length > NUM_OF_LINES);
  }, []);

  const handleRate = async(ratingComment, value) => {
    if (processing) return;
    setProcessing(true);
    const token = await AsyncStorage.getItem('authToken');
    const newComment = await rate(ratingComment, value, token);
    newComment && setProcessing(false);
  };

  const rate = async(ratingComment, value, token) => {
    if (ratingComment.ratings.current_user_rating_id) {
      if (ratingComment.ratings.current_user_rating_value !== value) {
        await API.changeRating(
          ratingComment.comment_content_type,
          ratingComment.id,
          ratingComment.ratings.current_user_rating_id,
          {value: value},
          token
        );
      }
      else {
        await API.changeRating(
          ratingComment.comment_content_type,
          ratingComment.id,
          ratingComment.ratings.current_user_rating_id,
          {value: 0},
          token
        );
      }
    }
    else {
      await API.postRating(ratingComment.comment_content_type, ratingComment.id, {value: value}, token);
    }
    return await fetchComment();
  };

  const fetchComment = () => {
    return AsyncStorage.getItem('authToken')
      .then((token) => API.getComment(comment.content_type, comment.object_pk, comment.id, token))
      .then(fetchedComment => {
        setComment(fetchedComment);
        return fetchedComment;
      });
  };

  const deleteComment = () => {
    AsyncStorage.getItem('authToken')
      .then((token) => API.deleteComment(commentBeingProcessed.content_type, commentBeingProcessed.object_pk, commentBeingProcessed.id, token))
      .then((response) => {
        const {statusCode, data} = response;
        props.toggleDeleteModal();
        if (statusCode == 200) {
          Alert.alert('Your comment was deleted.', 'Thank you for participating!',  [{ text: 'Ok' }]);
          fetchComment();
        }
        else {
          const errorMessage = 'That did not work.';
          let errorDetail;
          if (statusCode==403) {
            errorDetail = data.detail;
          }
          else if (statusCode == 400) {
            errorDetail = 'Bad request';
          }
          Alert.alert(errorMessage, errorDetail, [{ text: 'Ok' }]);
        }
      });
  };

  const getCommentTextDisplay = (comment) => {
    if (comment.is_removed) {
      return 'Deleted by creator on ' + DateService(comment.modified);
    }
    else if (comment.is_censored || comment.is_blocked) {
      return 'Deleted by moderation on '+ DateService(comment.modified);
    }
    else {
      return comment.comment;
    }
  };

  const isDisplayed = (comment) => {
    return !(comment.is_deleted || comment.is_blocked);
  };

  const optionsIcon = (<IconSLI name='options-vertical' size={22} />);
  const arrowUpIcon = (<IconSLI name='arrow-up' size={18} />);
  const arrowDownIcon = (<IconSLI name='arrow-down' size={18} />);
  const commentIcon = (<IconSLI name='bubble' size={18} />);
  const redoIcon = (<IconSLI name='action-redo' size={18} />);

  return (
    <View style={styles.container}>
      <View style={styles.top}>
        <View style={styles.topLeft}>
          {isDisplayed(comment) &&
          <Image source={{ uri: comment.user_image }} style={styles.avatar} />
          }
          <View style={styles.author}>
            <TextSourceSans style={styles.username}>{comment.user_name}</TextSourceSans>
            {isDisplayed(comment) &&
            <TextSourceSans style={styles.date}>{DateService(comment.created)}</TextSourceSans>
            }
          </View>
        </View>
        {isDisplayed(comment) &&
        <TextSourceSans>
          <Button
            icon={optionsIcon}
            type='clear'
            onPress={() => {setCommentBeingProcessed(comment); props.toggleMenu();}}
          />
        </TextSourceSans>
        }
      </View>
      <View>
        {!showWholeComment &&
          <TextSourceSans
            style={styles.comment}
            numberOfLines={NUM_OF_LINES}
            onTextLayout={onTextLayout}
          >
            {getCommentTextDisplay(comment)}
          </TextSourceSans>
        }
        {showWholeComment &&
        <TextSourceSans style={styles.comment}>
          {comment.comment}
        </TextSourceSans>
        }
        {hasExcerpt && <TouchableWithoutFeedback onPress={toggleWholeComment}>
          <TextSourceSans style={styles.linkButton}>{showWholeComment ? 'Read Less' : 'Read More'}</TextSourceSans>
        </TouchableWithoutFeedback>}
      </View>
      <View style={styles.linkSection}>
        {comment.child_comments.length !== 0 && <TouchableWithoutFeedback onPress={toggleSubComments}>
          <TextSourceSans style={styles.linkButton}>
            {showSubComments ? 'Hide' : 'Show'} {comment.child_comments.length} answers
          </TextSourceSans>
        </TouchableWithoutFeedback>}
      </View>
      <View style={styles.bottomActionsContainer}>
        <View style={styles.ratingButtons}>
          <ButtonCounter
            icon={arrowUpIcon}
            counter={comment.ratings.positive_ratings}
            onPress={() => handleRate(comment, 1)}
            highlight={
              comment.ratings.current_user_rating_id &&
              comment.ratings.current_user_rating_value === 1 &&
              comment.ratings.current_user_rating_value
            }
            disabled={!comment.has_rating_permission}
          />
          <ButtonCounter
            icon={arrowDownIcon}
            counter={comment.ratings.negative_ratings}
            onPress={() => handleRate(comment, -1)}
            highlight={
              comment.ratings.current_user_rating_id &&
              comment.ratings.current_user_rating_value === -1 &&
              comment.ratings.current_user_rating_value
            }
            disabled={!comment.has_rating_permission}
          />
        </View>
        <Button
          icon={commentIcon}
          title="Reply"
          titleStyle={styles.buttonTitle}
          type='clear'
          styles={styles.commentButton}
          onPress={() => {props.handleReply(comment.comment_content_type, comment.id);}}
          disabled={!props.hasCommentingPermission}
        />
        <Button
          icon={redoIcon}
          title="Share"
          titleStyle={styles.buttonTitle}
          type='clear'
        />
      </View>
      {showSubComments &&
        <SubComments
          comments={comment.child_comments}
          handleRate={handleRate}
          setCommentBeingProcessed={setCommentBeingProcessed}
          toggleMenu={props.toggleMenu}
          getCommentTextDisplay={getCommentTextDisplay}
          isDisplayed={isDisplayed}
        />
      }
    </View>
  );
};
