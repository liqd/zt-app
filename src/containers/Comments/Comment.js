import React, { useCallback, useEffect, useRef,useState } from 'react'
import { useTranslation } from 'react-i18next'
import { Alert, Image, View } from 'react-native'
import { TouchableWithoutFeedback } from 'react-native'
import IconSLI from 'react-native-vector-icons/SimpleLineIcons'
import { Button } from '@rneui/base'

import API from '../../BaseApi'
import { ButtonCounter } from '../../components/ButtonCounter'
import { TextSourceSans } from '../../components/TextSourceSans'

import { styles } from './Comment.styles'
import { SubComments } from './SubComments'

const NUM_OF_LINES = 2

export const Comment = (props) => {
  const { t } = useTranslation()
  const [showSubComments, setShowSubComments] = useState(false)
  const [showWholeComment, setShowWholeComment] = useState(false)
  const [hasExcerpt, setHasExcerpt] = useState(false)
  const [comment, setComment] = useState(props.comment)
  const processing = useRef(false)

  const getCommentMenuItems = (commentInstance) => {
    return [
      {
        title: t('Edit'),
        icon: 'pencil',
        action: () => props.toggleEditing(commentInstance),
        isFirst: true,
        isAllowed: commentInstance.user_info.has_changing_permission
      },
      {
        title: t('Delete'),
        icon: 'trash',
        action: () => props.toggleDeleteModal(),
        isAllowed: commentInstance.user_info.has_deleting_permission
      },
      {
        title: t('Report'),
        icon: 'flag',
        action: () => {
          props.toggleMenu()
          props.navigation.navigate('ReportCreateMessage', {
            content_type: commentInstance.comment_content_type,
            object_pk: commentInstance.id
          })
        },
        isFirst: !commentInstance.user_info.has_changing_permission && !commentInstance.user_info.has_deleting_permission,
        isLast: true,
        isAllowed: true
      },
      {
        title: t('Cancel'),
        action: () => props.toggleMenu(),
        isCancel: true,
        isAllowed: true
      },
    ]
  }

  const getCommentDeleteModalItems = (commentInstance) => {
    return [
      {
        title: [t('This comment will be deleted.'), t('This action cannot be undone.')],
        isText: true
      },
      {
        title: t('Delete'),
        action: () => deleteComment(commentInstance)
      },
      {
        title: t('Cancel'),
        action: () => props.toggleDeleteModal(),
        isCancel: true
      },
    ]
  }

  useEffect(() => {
    setComment(props.comment)
    props.openSubComments && setShowSubComments(true)
  }, [props.openSubComments, props.comment])

  const toggleSubComments = () => {
    setShowSubComments(!showSubComments)
  }

  const toggleWholeComment = () => {
    setShowWholeComment(!showWholeComment)
  }

  const onTextLayout = useCallback(e => {
    setHasExcerpt(e.nativeEvent.lines.length > NUM_OF_LINES)
  }, [])

  function handleOptions(commentInstance) {
    props.setDeleteModalItems(getCommentDeleteModalItems(commentInstance))
    props.setMenuItems(getCommentMenuItems(commentInstance))
    props.toggleMenu()
  }

  const handleRate = async(commentInstance, value) => {
    if (processing.current) return
    processing.current = true
    const newComment = await rate(commentInstance, value)
    if (newComment) {
      processing.current = false
    }
  }

  const rate = async(commentInstance, value) => {
    if (commentInstance.ratings.current_user_rating_id) {
      if (commentInstance.ratings.current_user_rating_value !== value) {
        await API.changeRating(
          commentInstance.comment_content_type,
          commentInstance.id,
          commentInstance.ratings.current_user_rating_id,
          {value: value}
        )
      } else {
        await API.changeRating(
          commentInstance.comment_content_type,
          commentInstance.id,
          commentInstance.ratings.current_user_rating_id,
          {value: 0}
        )
      }
    } else {
      await API.postRating(
        commentInstance.comment_content_type,
        commentInstance.id,
        {value: value}
      )
    }
    return await fetchComment()
  }

  const fetchComment = () => {
    return API.getComment(
      comment.content_type,
      comment.object_pk,
      comment.id,
    )
      .then(response => {
        setComment(response.data)
        return response.data
      })
  }

  const deleteComment = (commentInstance) => {
    API.deleteComment(
      commentInstance.content_type,
      commentInstance.object_pk,
      commentInstance.id,
    )
      .then((response) => {
        const {statusCode, data} = response
        props.toggleDeleteModal()
        if (statusCode == 200) {
          if (commentInstance.user_info.is_users_own_comment) {
            Alert.alert(
              t('Your comment was deleted.'),
              t('Thank you for participating!'),
              [{ text: 'Ok' }]
            )
          } else {
            Alert.alert(
              t('This comment was deleted.'),
              t('Thank you for moderating!'),
              [{ text: 'Ok' }]
            )
          }
          fetchComment()
        } else {
          const errorMessage = 'That did not work.'
          let errorDetail
          if (statusCode==403) {
            errorDetail = data.detail
          } else if (statusCode == 400) {
            errorDetail = 'Bad request'
          }
          Alert.alert(errorMessage, errorDetail, [{ text: 'Ok' }])
        }
      })
  }

  const getCommentTextDisplay = (commentInstance) => {
    if (commentInstance.is_removed) {
      return t('Deleted by creator on ') + commentInstance.modified
    } else if (commentInstance.is_censored || commentInstance.is_blocked) {
      return t('Deleted by moderation on ')+ commentInstance.modified
    } else {
      return commentInstance.comment
    }
  }

  const isDisplayed = (commentInstance) => {
    return !(commentInstance.is_deleted || commentInstance.is_blocked)
  }

  const optionsIcon = (<IconSLI name='options-vertical' size={22} />)
  const arrowUpIcon = (<IconSLI name='arrow-up' size={18} />)
  const arrowDownIcon = (<IconSLI name='arrow-down' size={18} />)
  const commentIcon = (<IconSLI name='bubble' size={18} />)
  const redoIcon = (<IconSLI name='action-redo' size={18} />)

  return (
    <View style={styles.container}>
      <View style={styles.top}>
        <View style={styles.topLeft}>
          {isDisplayed(comment) &&
          <Image
            source={{ uri: (comment.user_image) ? comment.user_image : comment.user_image_fallback }}
            style={styles.avatar}
            accessibilityIgnoresInvertColors={true}
          />
          }
          <View style={styles.author}>
            <TextSourceSans style={styles.username}>
              {comment.user_name}
            </TextSourceSans>
            {isDisplayed(comment) &&
            <TextSourceSans style={styles.date}>
              {comment.created}
            </TextSourceSans>
            }
          </View>
        </View>
        {isDisplayed(comment) &&
        <TextSourceSans>
          <Button
            testID={'options_button_' + comment.id}
            icon={optionsIcon}
            type='clear'
            onPress={() => handleOptions(comment)}
          />
        </TextSourceSans>
        }
      </View>
      <View>
        {!showWholeComment &&
          <TextSourceSans
            style={styles.comment}
            numberOfLines={NUM_OF_LINES}
            onTextLayout={onTextLayout}
          >
            {getCommentTextDisplay(comment)}
          </TextSourceSans>
        }
        {showWholeComment &&
        <TextSourceSans style={styles.comment}>
          {comment.comment}
        </TextSourceSans>
        }
        {hasExcerpt &&
          <TouchableWithoutFeedback
            accessibilityRole="button"
            onPress={toggleWholeComment}
          >
            <TextSourceSans style={styles.linkButton}>
              {showWholeComment ? t('Read Less') : t('Read More')}
            </TextSourceSans>
          </TouchableWithoutFeedback>}
      </View>
      <View style={styles.linkSection}>
        {comment.child_comments.length !== 0 &&
          <TouchableWithoutFeedback
            accessibilityRole="button"
            onPress={toggleSubComments}
          >
            <TextSourceSans style={styles.linkButton}>
              {t('{{ state }} {{ count }} answers',
                {state: showSubComments ? t('Hide') : t('Show') ,
                  count: comment.child_comments.length})}
            </TextSourceSans>
          </TouchableWithoutFeedback>
        }
      </View>
      <View style={styles.bottomActionsContainer}>
        <View style={styles.ratingButtons}>
          <ButtonCounter
            icon={arrowUpIcon}
            labelText={t('up-votes')}
            hintText={t('click to up vote')}
            counter={comment.ratings.positive_ratings}
            onPress={() => handleRate(comment, 1)}
            highlight={
              comment.ratings.current_user_rating_id &&
              comment.ratings.current_user_rating_value === 1 &&
              comment.ratings.current_user_rating_value
            }
            disabled={!comment.user_info.has_rating_permission}
          />
          <ButtonCounter
            icon={arrowDownIcon}
            labelText={t('down-votes')}
            hintText={t('click to down vote')}
            counter={comment.ratings.negative_ratings}
            onPress={() => handleRate(comment, -1)}
            highlight={
              comment.ratings.current_user_rating_id &&
              comment.ratings.current_user_rating_value === -1 &&
              comment.ratings.current_user_rating_value
            }
            disabled={!comment.user_info.has_rating_permission}
          />
        </View>
        <Button
          icon={commentIcon}
          title={t('Reply')}
          titleStyle={styles.buttonTitle}
          type='clear'
          styles={styles.commentButton}
          onPress={() => {props.handleReply(comment.comment_content_type, comment.id)}}
          disabled={!props.hasCommentingPermission}
        />
        <Button
          icon={redoIcon}
          title={t('Share')}
          titleStyle={styles.buttonTitle}
          type='clear'
        />
      </View>
      {showSubComments &&
        <SubComments
          comments={comment.child_comments}
          handleRate={handleRate}
          handleOptions={handleOptions}
          toggleMenu={props.toggleMenu}
          getCommentTextDisplay={getCommentTextDisplay}
          isDisplayed={isDisplayed}
        />
      }
    </View>
  )
}
