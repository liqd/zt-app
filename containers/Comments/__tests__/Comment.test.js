import React from 'react'
import { fireEvent, render } from '@testing-library/react-native'
import { Comment } from '../Comment'
import { testComment } from '../../../tests/TestData'

test('Test Comment Snapshot', () => {
  const { toJSON } = render(<Comment comment={testComment} />)
  expect(toJSON()).toMatchSnapshot()
})

test('Test Comment Toggle Menu', () => {
  let menuItems = null
  const toggleMenu = jest.fn()
  const setDeleteModalItems = jest.fn()
  const setMenuItems = jest.fn((m) => {
    menuItems = m
  })
  const { getByTestId } = render(
    <Comment
      comment={testComment}
      toggleMenu={toggleMenu}
      setMenuItems={setMenuItems}
      setDeleteModalItems={setDeleteModalItems}
    />)
  const menuButton = getByTestId('options_button_' + testComment.id)
  fireEvent.press(menuButton)
  expect(setDeleteModalItems).toBeCalledTimes(1)
  expect(setMenuItems).toBeCalledTimes(1)
  expect(menuItems).toHaveLength(4)
  expect(menuItems[0].title).toEqual('Edit')
  expect(menuItems[0].isAllowed).toBeFalsy()
  expect(menuItems[1].title).toEqual('Delete')
  expect(menuItems[1].isAllowed).toBeFalsy()
  expect(menuItems[2].title).toEqual('Report')
  expect(menuItems[2].isAllowed).toBeTruthy()
  expect(menuItems[2].isLast).toBeTruthy()
  expect(menuItems[3].title).toEqual('Cancel')
  expect(menuItems[3].isAllowed).toBeTruthy()
  expect(menuItems[3].isCancel).toBeTruthy()
  expect(toggleMenu).toBeCalledTimes(1)
})

test('Test Comment Toggle Menu has_changing_permission', () => {
  const canChangeComment = {
    ...testComment,
    user_info: { ...testComment.user_info, has_changing_permission: true }
  }
  let menuItems = null
  const toggleMenu = jest.fn()
  const setDeleteModalItems = jest.fn()
  const setMenuItems = jest.fn((m) => {
    menuItems = m
  })
  const { getByTestId } = render(
    <Comment
      comment={canChangeComment}
      toggleMenu={toggleMenu}
      setMenuItems={setMenuItems}
      setDeleteModalItems={setDeleteModalItems}
    />)
  const menuButton = getByTestId('options_button_' + testComment.id)
  fireEvent.press(menuButton)
  expect(setDeleteModalItems).toBeCalledTimes(1)
  expect(setMenuItems).toBeCalledTimes(1)
  expect(menuItems).toHaveLength(4)
  expect(menuItems[0].title).toEqual('Edit')
  expect(menuItems[0].isAllowed).toBeTruthy()
  expect(menuItems[1].title).toEqual('Delete')
  expect(menuItems[1].isAllowed).toBeFalsy()
  expect(menuItems[2].title).toEqual('Report')
  expect(menuItems[2].isAllowed).toBeTruthy()
  expect(menuItems[2].isLast).toBeTruthy()
  expect(menuItems[3].title).toEqual('Cancel')
  expect(menuItems[3].isAllowed).toBeTruthy()
  expect(menuItems[3].isCancel).toBeTruthy()
  expect(toggleMenu).toBeCalledTimes(1)
})

test('Test Comment Deleted', () => {
  const deletedComment = { ...testComment, is_deleted: true, is_removed: true }
  const { getByText, queryByText } = render(<Comment comment={deletedComment} />)
  expect(getByText(/Deleted by creator on/)).toBeTruthy()
  expect(queryByText(/January 1, 2022/)).toBeFalsy()
})

test('Test Comment Censored', () => {
  const censoredComment = { ...testComment, is_blocked: true, is_censored: true }
  const { getByText, queryByText } = render(<Comment comment={censoredComment} />)
  expect(getByText(/Deleted by moderation on/)).toBeTruthy()
  expect(queryByText(/January 1, 2022/)).toBeFalsy()
})
