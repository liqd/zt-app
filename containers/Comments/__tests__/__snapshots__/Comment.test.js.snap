// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test Comment Snapshot 1`] = `
<View
  style={
    Object {
      "borderColor": "#c4c4c4",
      "borderTopWidth": 1,
      "marginVertical": 10,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "flex-start",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 20,
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
          "flexDirection": "row",
        }
      }
    >
      <Image
        source={
          Object {
            "uri": "",
          }
        }
        style={
          Object {
            "backgroundColor": "#2d40cc",
            "borderRadius": 20,
            "height": 40,
            "marginRight": 20,
            "width": 40,
          }
        }
      />
      <View
        style={
          Object {
            "flex": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "fontWeight": "800",
              },
            ]
          }
        >
          test-user
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "color": "#666",
                "fontSize": 13.600000000000001,
              },
            ]
          }
        >
          January 1, 2022
        </Text>
      </View>
    </View>
    <Text
      style={
        Object {
          "fontFamily": "SourceSansPro_400Regular",
        }
      }
    >
      <View
        style={
          Array [
            Object {
              "overflow": "hidden",
            },
            Object {
              "borderRadius": 2,
            },
            undefined,
            false,
          ]
        }
        testID="RNE_BUTTON_WRAPPER"
      >
        <View
          accessibilityRole="button"
          accessibilityState={
            Object {
              "busy": false,
              "disabled": false,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "opacity": 1,
            }
          }
          testID="options_button_1"
        >
          <View
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "transparent",
                "borderColor": "#2089dc",
                "borderRadius": 2,
                "borderWidth": 0,
                "flexDirection": "row",
                "justifyContent": "center",
                "padding": 8,
                "paddingHorizontal": 10,
              }
            }
          >
            <Text />
          </View>
        </View>
      </View>
    </Text>
  </View>
  <View>
    <Text
      numberOfLines={2}
      onTextLayout={[Function]}
      style={
        Object {
          "fontFamily": "SourceSansPro_400Regular",
        }
      }
    >
      Lorem ipsum dolor sit amet, consetetur sadipscing elitr, sed diam nonumy eirmod tempor invidunt ut labore et dolore magna aliquyam
    </Text>
  </View>
  <View
    style={
      Object {
        "flexDirection": "row",
        "justifyContent": "space-between",
        "marginVertical": 20,
      }
    }
  />
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "row",
        "justifyContent": "space-between",
        "paddingBottom": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "width": 120,
        }
      }
    >
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "color": "#c4c4c4",
                "letterSpacing": 5,
              },
            ]
          }
        >
          0
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "color": "#c4c4c4",
                "fontSize": 20.4,
                "letterSpacing": 10,
                "lineHeight": 24.200000000000003,
              },
            ]
          }
        >
          <Text />
        </Text>
      </View>
      <View
        accessibilityState={
          Object {
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "flex": 1,
            "flexDirection": "row",
            "opacity": 1,
          }
        }
      >
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "color": "#c4c4c4",
                "letterSpacing": 5,
              },
            ]
          }
        >
          0
        </Text>
        <Text
          style={
            Array [
              Object {
                "fontFamily": "SourceSansPro_400Regular",
              },
              Object {
                "color": "#c4c4c4",
                "fontSize": 20.4,
                "letterSpacing": 10,
                "lineHeight": 24.200000000000003,
              },
            ]
          }
        >
          <Text />
        </Text>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": true,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#2089dc",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text />
          <Text
            style={
              Object {
                "color": "hsl(208, 8%, 63%)",
                "fontSize": 18,
                "marginLeft": 8,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Reply
          </Text>
        </View>
      </View>
    </View>
    <View
      style={
        Array [
          Object {
            "overflow": "hidden",
          },
          Object {
            "borderRadius": 2,
          },
          undefined,
          false,
        ]
      }
      testID="RNE_BUTTON_WRAPPER"
    >
      <View
        accessibilityRole="button"
        accessibilityState={
          Object {
            "busy": false,
            "disabled": false,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "transparent",
              "borderColor": "#2089dc",
              "borderRadius": 2,
              "borderWidth": 0,
              "flexDirection": "row",
              "justifyContent": "center",
              "padding": 8,
              "paddingHorizontal": 10,
            }
          }
        >
          <Text />
          <Text
            style={
              Object {
                "color": "#464a4c",
                "fontSize": 18,
                "marginLeft": 8,
                "paddingVertical": 1,
                "textAlign": "center",
              }
            }
          >
            Share
          </Text>
        </View>
      </View>
    </View>
  </View>
</View>
`;
